spec/event_builder_spec.rb

```diff
@@ -34,216 +34,336 @@
-      'decision' => {
-        'variationId' => '111128',
-        'experimentId' => '111127',
-        'isLayerHoldback' => false,
-      },
-      'layerId' => '1',
-      'accountId' => '12001',
-      'clientEngine' => 'ruby-sdk',
-      'clientVersion' => Optimizely::VERSION,
-      'userFeatures' => [],
+      account_id: '12001',
+      project_id: '111001',
+      visitors: [{
+        attributes: [],
+        visitor_id: 'test_user',
+        snapshots: [{
+          decisions: [{
+            campaign_id: '1',
+            experiment_id: '111127',
+            variation_id: '111128'
+          }],
+          events: [{
+            entity_id: '1',
+            timestamp: (time_now.to_f * 1000).to_i,
+            uuid: 'a68cf1ad-0393-4e18-af87-efe8f01a7c9c',
+            key: 'campaign_activated'
+          }]
+        }]
+      }],
+      revision: '42',
+      client_name: Optimizely::CLIENT_ENGINE,
+      client_version: Optimizely::VERSION
     }
-
-    @expected_conversion_url = 'https://logx.optimizely.com/log/event'
     @expected_conversion_params = {
-      'visitorId' => 'test_user',
-      'timestamp' => (time_now.to_f * 1000).to_i,
-      'isGlobalHoldback' => false,
-      'projectId' => '111001',
-      'accountId' => '12001',
-      'clientEngine' => 'ruby-sdk',
-      'clientVersion' => Optimizely::VERSION,
-      'userFeatures' => [],
-      'eventMetrics' => [],
-      'eventFeatures' => [],
-      'eventName' => 'test_event',
-      'eventEntityId' => '111095',
-      'layerStates' => [{
-        'layerId' => '1',
-        'decision' => {
-          'variationId' => '111128',
-          'experimentId' => '111127',
-          'isLayerHoldback' => false,
-        },
-        'actionTriggered' => true,
+      account_id: '12001',
+      project_id: '111001',
+      visitors: [{
+        attributes: [],
+        visitor_id: 'test_user',
+        snapshots: [{
+          decisions: [{
+            campaign_id: '1',
+            experiment_id: '111127',
+            variation_id: '111128'
+          }],
+          events: [{
+            entity_id: '111095',
+            timestamp: (time_now.to_f * 1000).to_i,
+            uuid: 'a68cf1ad-0393-4e18-af87-efe8f01a7c9c',
+            key: 'test_event'
+          }]
+        }]
       }],
+      revision: '42',
+      client_name: Optimizely::CLIENT_ENGINE,
+      client_version: Optimizely::VERSION
     }
+
   end
 
-  it 'should create a valid V2 Event when create_impression_event is called' do
+  it 'should create valid V2 Event when create_impression_event is called without attributes' do
     experiment = config.get_experiment_from_key('test_experiment')
     impression_event = @event_builder.create_impression_event(experiment, '111128', 'test_user', nil)
     expect(impression_event.params).to eq(@expected_impression_params)
-    expect(impression_event.url).to eq(@expected_impression_url)
+    expect(impression_event.url).to eq(@expected_event_builder_url)
```

Comment:
can we compare the whole object instead of properties inside impression_event?

Reply:
I suggest that either we construct whole object and compare it, or write a private function that does these comparisons for all tests: 
expect(impression_event.params).to eq(@expected_impression_params)
expect(impression_event.url).to eq(@expected_event_builder_url)
expect(impression_event.http_verb).to eq(:post)
Add also the headers

Reply:
@msohailhussain
I have updated these tests as per changes.
Each value is tested separately which is more desirable and it help pin point the issue is any case fails. I would not suggest removing it in favor of object matching.